File: src/java/DutchNationalFlag.java
public class DutchNationalFlag {

    private void swap(int[] a, int index1, int index2) {
        int temp = a[index1];
        a[index1] = a[index2];
        a[index2] = temp;
    }

    public void dutchNationalFlag(int[] a, int n) {
        //红色游标
        int left = 0;
        //蓝色游标
        int right = n - 1;
        //白色游标
        int cur = 1;
        while (cur <= right) {
            if (a[cur] == 0) { //红色
                swap(a, left, cur);
                left++;
            } else if (a[cur] == 1) { //白色
                cur++;
            } else { //蓝色
                swap(a, cur, right);
                right--;
            }
        }
    }

    public static void main(String[] args) {
        int[] a = { 0, 1, 1, 2, 2, 2, 0, 0, 0, 1, 1 };
        DutchNationalFlag dFlag = new DutchNationalFlag();
        dFlag.dutchNationalFlag(a, a.length);
        for (int value : a) {
            System.out.print(value);
            System.out.print(" ");
        }
    }
}

File: src/javascript/dutch-national-flag.js
function swap(arr, i, j) {
  let temp = arr[i];
  arr[i] = arr[j];
  arr[j] = temp;
}

function dutchNationalFlag(arr) {
  if (arr.length < 2) {
    return;
  }
  let left = 0;
  let cur = 1;
  let right = arr.length - 1;
  while (cur <= right) {
    if (arr[cur] == 0) {
      swap(arr, cur, left);
      left++;
    } else if (arr[cur] == 1) {
      cur++;
    } else {
      swap(arr, cur, right);
      right--;
    }
  }
}

var arr = [2, 1, 0, 0, 0, 2, 2, 1, 1];
dutchNationalFlag(arr);
console.log(arr);